File: nreduce/benchmark/quicksort.java
import java.util.*;

public class quicksort
{
  static List<String> genlist(int max)
  {
    List<String> res = new ArrayList<String>();
    for (int i = 0; i < max; i++) {
      char[] chars = ((7*i)+"").toCharArray();
      char[] reversed = new char[chars.length];
      for (int c = 0; c < chars.length; c++)
        reversed[chars.length-1-c] = chars[c];
      res.add(new String(reversed));
    }
    return res;
  }

  static void printList(List<String> list)
  {
    for (String s : list)
      System.out.println(s);
  }

  static List<String> quicksort(List<String> list)
  {
    String pivot = list.get(0);
    List<String> before = new ArrayList<String>();
    List<String> after = new ArrayList<String>();
    for (int pos = 1; pos < list.size(); pos++) {
      String cur = list.get(pos);
      if (cur.compareTo(pivot) < 0)
        before.add(cur);
      else
        after.add(cur);
    }
    List<String> res = new ArrayList<String>();
    if (!before.isEmpty())
      res.addAll(quicksort(before));
    res.add(pivot);
    if (!after.isEmpty())
      res.addAll(quicksort(after));
    return res;
  }

  public static void main(String[] args)
  {
    int n = 1000;
    if (args.length > 0)
      n = Integer.parseInt(args[0]);
    List<String> input = genlist(n);
    List<String> sorted = quicksort(input);
    printList(sorted);
  }
}

File: nreduce/benchmark/quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

char **genlist(int max)
{
  char **res = (char**)malloc(max*sizeof(char*));
  int i;
  for (i = 0; i < max; i++) {
    char tmp[100];
    char *rev;
    int len;
    int c;
    sprintf(tmp,"%d",7*i);
    len = strlen(tmp);
    rev = (char*)malloc(len+1);
    for (c = 0; c < len; c++)
      rev[c] = tmp[len-1-c];
    rev[len] = '\0';
    res[i] = rev;
  }
  return res;
}

char **quicksort(char **list, int len)
{
  char *pivot = list[0];
  char **res = (char**)malloc(len*sizeof(char*));
  char **before = (char**)malloc(len*sizeof(char*));
  char **after = (char**)malloc(len*sizeof(char*));
  int nbefore = 0;
  int nafter = 0;
  int pos;

  for (pos = 1; pos < len; pos++) {
    char *cur = list[pos];
    if (strcmp(cur,pivot) < 0)
      before[nbefore++] = cur;
    else
      after[nafter++] = cur;
  }

  if (0 < nbefore) {
    char **sbefore = quicksort(before,nbefore);
    memcpy(&res[0],sbefore,nbefore*sizeof(char*));
    free(sbefore);
  }
  res[nbefore] = pivot;
  if (0 < nafter) {
    char **safter = quicksort(after,nafter);
    memcpy(&res[nbefore+1],safter,nafter*sizeof(char*));
    free(safter);
  }

  return res;
}

int main(int argc, char **argv)
{
  int n = 1000;
  char **strings;
  char **sorted;
  int i;
  if (2 <= argc)
    n = atoi(argv[1]);

  strings = genlist(n);

  sorted = quicksort(strings,n);

  for (i = 0; i < n; i++)
    printf("%s\n",sorted[i]);

  for (i = 0; i < n; i++)
    free(strings[i]);
  free(strings);
  free(sorted);
  return 0;
}
